File: pickr/lib/classes/games/game.dart
import 'package:flutter/material.dart';

import 'package:pickr/classes/card.dart';
import 'package:pickr/classes/deck.dart';
import 'package:pickr/classes/player.dart';

abstract class Game {
  //

  int numplayers;
  int score;
  int _round = 1;
  Deck _deck;
  List<GamingCard> _table;
  List<Player> _players;

  Game({@required int numplayers, @required int score}) {
    numplayers = numplayers;
    score = score;
    _deck = Deck();
    _table = <GamingCard>[];
    _players = <Player>[];
  }

  /// Getter method for [_deck].
  Deck get deck => _deck;

  /// Getter method for [_round].
  int get round => _round;

  /// Getter method for [_table].
  List<GamingCard> get table => _table;

  /// Getter method for [_players].
  List<Player> get players => _players;

  /// Getter method for [_cards].
  List<GamingCard> get cards => _deck.cards;

  /// Distributes the initial cards to the players.
  void spread();

  /// Rule that determines the winner of the round.
  GamingCard rule();

  /// Adds a player to the [_players] list.
  void addPlayer({String id});

  /// Returns the amount of points of the cards on the table.
  int getTableValue() {
    var result = 0;
    table.forEach((card) {
      result += card.val;
    });
    return result;
  }

  /// Drops a card on the table
  void dropCard({GamingCard card}) => _table.add(card);

  /// Starts the game
  void start();

  /// Sequence of actions in an entire round of game.
  void gameround();

  /// Determines the winner of the round and awards him with
  /// the points it deserves.
  Player winner(GamingCard winnerCard) {
    var winner = players.first;

    _players.forEach((player) {
      if (player.hand.contains(winnerCard)) winner = player;

      player.delete();
    });

    winner.setGameScore = getTableValue();

    table.clear();

    _round += 1;

    return winner;
  }
}

File: pickr-java/src/pickr/games/Game.java
package pickr.games;


import java.util.ArrayList;

import pickr.classes.Deck;
import pickr.classes.GamingCard;
import pickr.classes.Player;
import pickr.exceptions.OutOfBoundException;
import pickr.exceptions.WrongCardNumberException;

public abstract class Game {
	
	int numplayers;
	int score;
	private int round = 0;
	protected Deck deck;
	private ArrayList<GamingCard> table;
	protected ArrayList<Player> players;
	
	public Game(int numplayers, int score) {
		this.numplayers = numplayers;
		this.score = score;
		this.deck = new Deck();
		this.table = new ArrayList<GamingCard>();
		this.players = new ArrayList<Player>();
	}

	public Deck getDeck() { return deck; }

	public int getRound() { return round; }

	public ArrayList<GamingCard> getTable() { return table; }

	public ArrayList<Player> getPlayers() { return players; }
	
	abstract void spread() throws OutOfBoundException;
	
	abstract GamingCard rule();
	
	public void addPlayer(String id) throws OutOfBoundException {
		if(players.size() < this.numplayers) {
			players.add(new Player(id));
		}
		else throw new OutOfBoundException();
	}
	
	int getTableValue() {
		int result = 0;
		for(GamingCard card : this.table) {
			result += card.getVal();
		}
		return result;
	}
	
	void dropCard(GamingCard card) { table.add(card); }
		
	abstract void start() throws WrongCardNumberException, OutOfBoundException;	
	
	public Player winner(GamingCard winnerCard) {		
		
		Player winner = this.players.get(0);
		
		for(Player player : this.players) {
				
			if(player.getCards().contains(winnerCard)) winner = player;
			
			player.delete();
		}
		
		System.out.println("the winner is " + winner.getId());	
		
		winner.setGameScore(getTableValue());
		
		table.clear();
		
		this.round += 1;
		
		return winner;		
	}

}
